tcp-server/src/main.rs

```diff
@@ -0,0 +1,116 @@
+use std::borrow::BorrowMut;
+use std::collections::vec_deque::Iter;
+use std::collections::VecDeque;
+use std::io;
+use std::io::{BufRead, BufReader, Read, Write};
+use std::net::{TcpListener, TcpStream};
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::thread::JoinHandle;
+use std::time::Duration;
+
+use redisish::{Command, parse};
+
+trait Mailbox {
+    /// TODO Is it fine that mutability is hidden? How can we do it some other way?
+    fn append(&self, email: String);
+    /// TODO does it make sense to return an abstract iterator and how can this be achieved?
+    fn list_emails(&self) -> String;
+}
+
+struct VecDequeMailbox {
+    data: Mutex<VecDeque<String>>,
+}
+
+impl Mailbox for VecDequeMailbox {
+    /// TODO is there a way (and does it make sense) to accept &str?
+    fn append(&self, email: String) {
+        let result = self.data.lock();
+        result.unwrap().push_front(email);
+    }
+
+    /// Returns a list of emails as a string, separated by `;`
+    fn list_emails(&self) -> String {
+        self.data.lock()
+            .unwrap()
+            .iter()
+            .fold(String::new(), |acc, next| acc + next + ";")
+    }
+}
+
+impl VecDequeMailbox {
+    // TODO how to return Box<dyn Mailbox>?
+    fn new() -> Box<VecDequeMailbox> {
+        return Box::new(VecDequeMailbox {
+            data: Mutex::new(VecDeque::new()),
+        });
+    }
+}
+
+/// General questions:
+/// * how to work with `dyn Mailbox ` in multithreaded environment?
+fn main() -> io::Result<()> {
+    spawn_tcp_listener();
+    let client_thread = spawn_monitoring_thread();
+    client_thread.join().unwrap();
+    Ok(())
+}
+
+/// Spawns TcpListener thread, connects to 127.0.0.1:8080 and spawns thread per connection
+fn spawn_tcp_listener() {
+    thread::spawn(|| {
+        let mailbox = Arc::new(VecDequeMailbox::new());
+        let listener = TcpListener::bind("127.0.0.1:8080").unwrap();
+        for stream in listener.incoming() {
+            // TODO is shadowing the variable name idiomatic?
+            let mailbox = mailbox.clone();
+            // TODO: confirm that dropping the JoinHandle does not do anything
```

Comment:
thread handles should be joined for proper error handling but it's not a fatal error to just drop them. In this case you can't wait for them to finish anymore, so any chance of re-sync is lost. See [JoinHandle docs](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html)